logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java

```diff
@@ -0,0 +1,345 @@
+            Graph otherG = (Graph) sourceComponent;
+            if (otherG.getVertices().size() != this.getVertices().size()) return false;
+
+            boolean edgesEqual = this.getEdges().stream().
+                    allMatch(e -> otherG.getEdges().stream().anyMatch(oe -> oe.sourceComponentEquals(e)));
+
+            // We need to check vertices separately because there may be unconnected vertices
+            boolean verticesEqual = this.getVertices().stream().
+                    allMatch(v -> otherG.getVertices().stream().anyMatch(ov -> ov.sourceComponentEquals(v)));
+
+            return edgesEqual && verticesEqual;
+        }
+        return false;
+    }
+
+    // returns true if this graph has a .sourceComponentEquals equivalent edge
+    public boolean hasEquivalentEdge(Edge otherE) {
+        return this.getEdges().stream().anyMatch(e -> e.sourceComponentEquals(otherE));
+    }
+
+
+    public class DiffResult {
+        public Collection<Edge> getRemovedEdges() {
+            return removedEdges;
+        }
+
+        public Collection<Edge> getAddedEdges() {
+            return addedEdges;
+        }
+
+        private final Collection<Edge> removedEdges;
+        private final Collection<Edge> addedEdges;
+
+        public DiffResult(Collection<Edge> removed, Collection<Edge> added) {
+            this.removedEdges = removed;
+            this.addedEdges = added;
+        }
+
+        public String toString() {
+            return "Diff Result (-" + removedEdges.size() + ",+" + addedEdges.size() + ")\n" +
+                    removedEdges.stream().map(e -> "-" + e.toString()).collect(Collectors.joining("\n")) +
+                    "\n" +
+                    addedEdges.stream().map(e -> "+" + e.toString()).collect(Collectors.joining("\n"));
+        }
+    }
+
+    public DiffResult diff(Graph o) {
+       List<Edge> removedEdges = this.getEdges().stream().filter(e -> !o.hasEquivalentEdge(e)).collect(Collectors.toList());
+       List<Edge> addedEdges = o.getEdges().stream().filter(e -> !this.hasEquivalentEdge(e)).collect(Collectors.toList());
+        return new DiffResult(removedEdges, addedEdges);
+    }
+
+    @Override
+    public SourceMetadata getMeta() {
+        return null;
+    }
+
+    public boolean isEmpty() {
+        return (this.getVertices().size() == 0);
+    }
+
+    public Graph threadToGraph(BooleanEdge.BooleanEdgeFactory edgeFactory, Vertex v, Graph otherGraph) throws InvalidIRException {
+        if (otherGraph.getVertices().size() == 0) return this;
+
+        for (Vertex otherRoot : otherGraph.getRoots()) {
+            this.threadVertices(edgeFactory, v, otherRoot);
+        }
+
+        return this;
+    }
+
+    // Return plugin vertices by type
+    public List<PluginVertex> getPluginVertices(PluginDefinition.Type type) {
+       return pluginVertices()
+               .filter(v -> v.getPluginDefinition().getType().equals(type))
+               .collect(Collectors.toList());
+    }
```

Comment:
Yes! I'm (trying) to follow the JDK8 convention of `getFoos -> Collection` and `foos -> Stream`. This is in violation of that!